Fixed #30231 -- Fixed admin filter horizontal/vertical verbose_name generation.

smithdc1 opened the pull request:
[Ticket #30231](https://code.djangoproject.com/ticket/30231), and follows #11282 

I've added three more tests to the previous version of this patch. This gives a total of 4 tests for the 4 changes in the JS. As the tests are identical I've used `subtest` to save a significant amount of repetition. 

The comments on the previous PR suggested exploring using an existing model. I've managed to do this but needed to change one of the existing models slightly while not breaking any of the existing tests. 

Hopefully, less typo's today 🙈

smithdc1 commented on tests/admin_inlines/tests.py:
```diff
@@ -1348,3 +1348,60 @@ def test_inline_formset_error(self):
+                    '.dynamic-fashionista_set')[0].get_attribute('id'), 'fashionista_set-0')
+                self.assertEqual(len(self.selenium.find_elements_by_css_selector(
+                    '.dynamic-fashionista_set#fashionista_set-0 h2')), 2)
+                self.assertEqual(self.selenium.find_elements_by_css_selector(
+                    '.dynamic-fashionista_set#fashionista_set-0 h2')[0].text, 'AVAILABLE LIMITATIONS')
+                self.assertEqual(self.selenium.find_elements_by_css_selector(
+                    '.dynamic-fashionista_set#fashionista_set-0 h2')[1].text, 'CHOSEN LIMITATIONS')
```
Do we need all of the test the whole structure? Could we 'just' test these two items? 🤔

smithdc1 commented:
buildbot, test on selenium.

jdufresne commented on django/contrib/admin/static/admin/js/inlines.js:
```diff
@@ -218,12 +218,10 @@
                 $('.selectfilter').each(function(index, value) {
-                    const namearr = value.name.split('-');
-                    SelectFilter.init(value.id, namearr[namearr.length - 1], false);
+                    SelectFilter.init(value.id, $(this).data('fieldName'), false);
```
Accessing the data attributes doesn't require jQuery. There have been some efforts to reduce jQuery usage in the admin, so can take advantage of that here.

```suggestion
                    SelectFilter.init(value.id, this.dataset.fieldName, false);
```

And the other lines too:

```diff
diff --git a/django/contrib/admin/static/admin/js/inlines.js b/django/contrib/admin/static/admin/js/inlines.js
index 777b76e1fe..d9a9032c45 100644
--- a/django/contrib/admin/static/admin/js/inlines.js
+++ b/django/contrib/admin/static/admin/js/inlines.js
@@ -218,10 +218,10 @@
             // instantiate a new SelectFilter instance for it.
             if (typeof SelectFilter !== 'undefined') {
                 $('.selectfilter').each(function(index, value) {
-                    SelectFilter.init(value.id, $(this).data('fieldName'), false);
+                    SelectFilter.init(value.id, this.dataset.fieldName, false);
                 });
                 $('.selectfilterstacked').each(function(index, value) {
-                    SelectFilter.init(value.id, $(this).data('fieldName'), true);
+                    SelectFilter.init(value.id, this.dataset.fieldName, true);
                 });
             }
         };
@@ -281,10 +281,10 @@
             // If any SelectFilter widgets were added, instantiate a new instance.
             if (typeof SelectFilter !== "undefined") {
                 $(".selectfilter").each(function(index, value) {
-                    SelectFilter.init(value.id, $(this).data('fieldName'), false);
+                    SelectFilter.init(value.id, this.dataset.fieldName, false);
                 });
                 $(".selectfilterstacked").each(function(index, value) {
-                    SelectFilter.init(value.id, $(this).data('fieldName'), true);
+                    SelectFilter.init(value.id, this.dataset.fieldName, true);
                 });
             }
         };
```

smithdc1 commented:
buildbot, test on selenium.

carltongibson commented on tests/admin_inlines/models.py:
```diff
@@ -160,8 +160,8 @@ class OutfitItem(models.Model):
 class Fashionista(models.Model):
-    person = models.OneToOneField(Person, models.CASCADE, primary_key=True)
-    weaknesses = models.ManyToManyField(OutfitItem, through='ShoppingWeakness', blank=True)
+    person = models.ForeignKey(Person, models.CASCADE)
```
It might be that we can't do this. See main comment.

smithdc1 replied:
I couldn't find a model that is suitable. There are only two models with `ManyToManyField` and both would need changing for these tests. I don't think the added effort of checking these changes is worth the effort.

carltongibson reviewed (commented):
Hi @smithdc1. 

I like the proxy model approach, but I'm not sure we can change `Fasionista` like this — or at least we need to double check. 
The model was introduced in dd07c235454baea0c74fa373cc07187942f3aed0 for ticket-12749. Description there begins...

> When a Model has a field with primary_key=True and the Admin has an Inline ...

i.e. If we remove the `primary_key` there do we still have a reproduce/regression test for the original issue? 🧐

It might be that we just need to add the new models. (Thought: It seems like there's a lot of them… **surely** one of 'em would work — but perhaps not. 😀 — Don't spend forever here!) 

I think we can squash the second commit in. We're already touching the lines in question.

smithdc1 commented:
Hi @carltongibson 

> > When a Model has a field with primary_key=True and the Admin has an Inline ...
> 
> i.e. If we remove the `primary_key` there do we still have a reproduce/regression test for the original issue? 🧐
> 
> It might be that we just need to add the new models. (Thought: It seems like there's a lot of them… **surely** one of 'em would work — but perhaps not. 😀 — Don't spend forever here!)

This was the best I could come up with on the existing models, but I'll have another look. It's a good call on changing the model. My observation & learning here is that I need to be really aware of the knock-on impact of changes, and the work it causes for others. That is, the ROI of checking the impact of a model change is likely not worth it vs 'just' adding another model if there isn't a straight forward solution.

smithdc1 commented:
buildbot, test on selenium.

smithdc1 commented on tests/admin_inlines/tests.py:
```diff
@@ -1348,3 +1348,37 @@ def test_inline_formset_error(self):
+                self.selenium.get(self.live_server_url + reverse(param))
+
+                # First inline shows the verbose_name
+                available, chosen = self.selenium.find_elements_by_css_selector(css_selector % 0)
```
Fomatting the `css_selector` in this way is a lot shorter, but I'm not so sure on readability. Part of me likes it as the difference is more clear as it avoids it being buried deep inside the string.

smithdc1 commented:
Ok -- I think this is ready for another review. I couldn't find a suitable model that exists so I've reverted to the standalone models, and stripped them back to the must have items only. I've also updated the tests, I hope these are slightly more readable while still maintaining a similar level of test coverage.

carltongibson commented:
buildbot, test on selenium.

carltongibson reviewed (approved):
Hey @smithdc1 — thanks for picking this up. Looks good. 

I've rebased, squashed, and pushed cosmetic edits. We'll just wait for CI but 👍 

@TeresaPartidaS thanks for the original effort writing the selenium test here — those are always a hurdle! Sorry again for the slow pickup my end. Welcome aboard ⛵

carltongibson commented:
buildbot, test on selenium.

felixxm commented:
@smithdc1 @carltongibson  Added tests crash in the `headless` mode:
```
Traceback (most recent call last):
  File "/home/jenkins/workspace/django-selenium/database/sqlite3/label/bionic/python/python3.9/tests/admin_inlines/tests.py", line 1382, in test_inlines_verbose_name
    self.assertEqual(chosen.text, 'CHOSEN ATTENDANT')
AssertionError: '' != 'CHOSEN ATTENDANT'
+ CHOSEN ATTENDANT
```

It's probably a separate CSS issue on small screens (800px), see screenshots:

- headless:
  ![screenshot_admin:admin_inlines_course_add](https://user-images.githubusercontent.com/2865885/108352590-5c9b5580-71e7-11eb-88df-0f267f3b4737.png)
- "normal":
  ![screenshot_admin:admin_inlines_course_add](https://user-images.githubusercontent.com/2865885/108352481-34135b80-71e7-11eb-84a3-66f7e8e3026a.png)

smithdc1 commented:
Looks like a regression following the introduction of the sidebar.

I'll write up more complete notes into a new ticket later (I'm not sure what a fix would look like) and submit a patch to fix the current failing test.

I'm also really not sure what I missed here, I thought I had seen the selenium tests passing? What did I get wrong here?

felixxm commented:
> I'll write up more complete notes into a new ticket later (I'm not sure what a fix would look like) and submit a patch to fix the current failing test.

Thanks :+1: 

> I'm also really not sure what I missed here, I thought I had seen the selenium tests passing? What did I get wrong here?

Nothing. We run selenium tests in headless mode only in the main build (see [django-selenium](https://djangoci.com/view/Main/job/django-selenium/)) not for every PR (see [pull-requests-selenium](https://djangoci.com/view/PR%20Builders/job/pull-requests-selenium/)).
